### PR TITLE
Implement Valtree to ConstValue conversion

### PR DESCRIPTION
**Pull request by b-naber**
Once we start to use `ValTree`s in the type system we will need to be able to convert them into `ConstValue` instances, which we want to continue to use after MIR construction.

r? @oli-obk 

cc @RalfJung

### CODE REVIEW

**Comment by RalfJung** [comment]
I am surprised by `ConstValue` as the conversion target though. Isn't the plan to entirely get rid of `ConstValue`? Those should all become valtrees.

**Comment by b-naber** [comment]
> I am surprised by `ConstValue` as the conversion target though. Isn't the plan to entirely get rid of `ConstValue`? Those should all become valtrees.

@RalfJung We were debating this [here](https://rust-lang.zulipchat.com/#narrow/stream/260443-project-const-generics/topic/valtrees/near/269732118).

**Comment by RalfJung** [comment]
That doesn't really mention the point that ConstValue are supposed to be removed.

@lcnr wrote
> would have to check whether we would then have to reimplement the conversion for each backend

However, I think if valtrees can be converted into a `ConstAlloc`, then that's enough for all backends? Most backends will *probably* want to have a shortcut for certain small valtrees (like, of integer type), but I don't think we want to keep the ConstValue type around.

**Comment by oli-obk** [comment]
For now, moving to just ConstAlloc instead of ConstValue is a too big change to be in one step. We can either dynamically forbid it in MIR and convert everywhere to ConstAlloc as a test, or we can do this PR and then explore that possibility afterwards. I'm slightly in favour of the latter, as the scope of when to convert to ConstAlloc is smaller then

**Comment by lcnr** [comment]
can miri also directly work with `ConstAlloc`? That conversion is necessary because we have to go from `ty::Const` to `mir::ConstantKind` for ctfe

**Comment by b-naber** [comment]
Wouldn't there be problems in the query system if we would convert `ValTree`s into `ConstAlloc`s? Is the idea to keep the `AllocId`s stored somewhere and then map a `ValTree` to an `AllocId`? Otherwise if we would create a new `Allocation` for a `ValTree` in the conversion, afaict that would basically amount to most of what this PR is doing.

**Comment by RalfJung** [comment]
`ConstAlloc` should be trivial to use for Miri, much easier than any other representation.

But if this is just a stepping stone, and the plan is still to remove ConstValue later, and most of the work done here will be relevant even after ConstValue removal, then I am happy. :)

**Comment by lcnr** [comment]
yes, I still don't think we should cache the origin for valtrees so that we don't have to rebuild them later, that's prone to cause unstable query results, leading to ICE or unsoudness

**Comment by b-naber** [comment]
@RalfJung Since Oli is on vacation, can you maybe review this, please? This is only CTFE related code.

**Comment by oli-obk** [comment]
CI is failing... maybe we should teach CI to keep compiling on lint errors but fail only at the end

File: compiler/rustc_const_eval/src/interpret/place.rs [hunk 1]
```diff
@@ -100,7 +100,7 @@ impl<'tcx, Tag: Provenance> std::ops::Deref for PlaceTy<'tcx, Tag> {
 /// A MemPlace with its layout. Constructing it is only possible in this module.
 #[derive(Copy, Clone, Hash, Eq, PartialEq, Debug)]
 pub struct MPlaceTy<'tcx, Tag: Provenance = AllocId> {
-    mplace: MemPlace<Tag>,
+    pub(crate) mplace: MemPlace<Tag>,
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/interpret/place.rs:103]: Keep it private

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+                        create_mplace_from_layout(ecx, ty::ParamEnv::empty().and(arr_ty));
+                    debug!(?arr_place);
+
+                    // Insert elements of `arr_valtree` into `place`
+                    fill_place_recursively(ecx, &mut arr_place, valtree, arr_ty);
+                    dump_place(&ecx, arr_place.into());
+
+                    // Now we need to create a `ScalarPair` from the filled `place`
+                    // and write that into `place`
+                    let (alloc_id, offset) = arr_place.mplace.ptr.into_parts();
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:348]: Instead of taking the pointer apart and putting it back together, just use the pointer directly.

Also use the deref impl to get the field instead of accessing the field

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+        | ty::Closure(..)
+        | ty::Generator(..)
+        | ty::GeneratorWitness(..) => None,
+    }
+}
+
+#[instrument(skip(ecx), level = "debug")]
+fn create_mplace_from_layout<'tcx>(
+    ecx: &mut CompileTimeEvalContext<'tcx, 'tcx>,
+    param_env_ty: ty::ParamEnvAnd<'tcx, Ty<'tcx>>,
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:142]: Grab the param env from the ecx  to avoid passing it at allcall sites

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+                    let data = match tcx.get_global_alloc(alloc_id) {
+                        Some(GlobalAlloc::Memory(const_alloc)) => const_alloc,
+                        _ => bug!("expected memory allocation"),
+                    };
+                    debug!(?data);
+
+                    return ConstValue::Slice { data, start: 0, end: len as usize };
+                }
+                _ => {
+                    match valtree {
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:215]: Except for the interning, these arms are exactly the same. Why don't we need interning in the second branch? Please deduplicate these two branches

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+    }
+}
+
+// FIXME Needs a better/correct name
+#[instrument(skip(ecx), level = "debug")]
+fn fill_place_recursively<'tcx>(
+    ecx: &mut CompileTimeEvalContext<'tcx, 'tcx>,
+    place: &mut MPlaceTy<'tcx>,
+    valtree: ty::ValTree<'tcx>,
+    ty: Ty<'tcx>,
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:304]: Place already has a type in the layout, no need to pass twice

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+                        .unwrap();
+
+                    let imm = pointee_place.mplace.to_ref(&tcx);
+                    debug!(?imm);
+
+                    ecx.write_immediate(imm, &(*place).into()).unwrap();
+                }
+            }
+        }
+        ty::Tuple(tuple_types) => {
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:382]: This and all the arms below can use the same logic (with just an adjustment to skip the variant logic and downcast for non-enums).

The field types are already in the mplace from the field projection, so you can ignore the types and not deconstruct them yourself

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+            debug!("writing trivial valtree {:?} to place {:?}", scalar_int, place);
+            ecx.write_immediate(
+                Immediate::Scalar(ScalarMaybeUninit::Scalar(scalar_int.into())),
+                &(*place).into(),
+            )
+            .unwrap();
+        }
+        ty::Ref(_, inner_ty, _) => {
+            match inner_ty.kind() {
+                ty::Slice(_) | ty::Str => {
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:323]: The fallback arm should mostly do the right thing for slices, too. So you should be able to deduplicate them and just add a small adjustment for the metadata within the logic

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_const_eval/src/interpret/place.rs [hunk 8]
```diff
@@ -617,19 +621,23 @@ where
 
     /// Computes a place. You should only use this if you intend to write into this
     /// place; for reading, a more efficient alternative is `eval_place_for_read`.
+    #[instrument(skip(self), level = "debug")]
     pub fn eval_place(
         &mut self,
         place: mir::Place<'tcx>,
     ) -> InterpResult<'tcx, PlaceTy<'tcx, M::PointerTag>> {
+        debug!("projection: {:?}", place.projection);
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/place.rs:629]: Please don't add tons of `debug!`. That makes it extremely hard to debug Miri since one just drowns in useless information.

If you must commit them, please make them the lowest verbosity level, `trace!`.

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs (hunk 2 above)
Comment (b-naber) [reply compiler/rustc_const_eval/src/const_eval/valtrees.rs:348]: The destructuring was unfortunately necessary since the pointer corresponding to the place has type `Pointer<Option<AllocId>>`, but `ScalarMaybeUninit::from_pointer` requires `Pointer<AllocId>>` (as does the deref call). Didn't find any way to unwrap the inner alloc_id.

Comment (b-naber) [reply compiler/rustc_const_eval/src/const_eval/valtrees.rs:348]: This was also why I had to make `offset` of `Pointer` non-private (which is supposed to be private). There has to be a way to do this, since I would assume `Pointer<Option<AllocId>>` is encountered quite often.

Comment (RalfJung) [reply compiler/rustc_const_eval/src/const_eval/valtrees.rs:348]: There is `ScalarMaybeUninit::from_maybe_pointer` for this exact case. :)

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+                let mut place_inner = match *ty.kind() {
+                    ty::Adt(def, substs) if !def.is_enum() => {
+                        let field = &def.variant(VariantIdx::from_usize(0)).fields[i];
+                        let field_ty = field.ty(tcx, substs);
+                        let projection_elem = ProjectionElem::Field(Field::from_usize(i), field_ty);
+
+                        ecx.mplace_projection(&place_adjusted, projection_elem).unwrap()
+                    }
+                    ty::Adt(_, _) | ty::Tuple(_) => ecx.mplace_field(&place_adjusted, i).unwrap(),
+                    ty::Array(_, _) | ty::Str => {
```
Comment (b-naber) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:333]: This still fails. I'm currently getting the following when compiling std during the conversion for a valtree with underlying type `&[&str; 1]`. In the recursive call for `str` we get:

```
│ │ │ │ │ ├─┐rustc_const_eval::const_eval::valtrees::fill_place_recursively place=MPlaceTy { mplace: MemPlace { ptr: alloc1411, align: Align { pow2: 0 }, meta: Meta(0x000000000000002a) }, layout: TyAndLayout { ty: str, layout: Layout { fields: Array { stride: Size { raw: 1 }, count: 0 }, variants: Single { index: 0 }, abi: Aggregate { sized: false }, largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 0 } }, size: Size { raw: 0 } } } }
``` 

(Had to manually add the meta data for the pointer in `create_mplace_from_layout`, which seems fishy, but I can't call deref_operand since the memory is uninit at this stage).

As a comparison in the valtree creation we have this call to `slice_branches` (which works):

```
│ │ │ │ │ │ │ │ ├─┐rustc_const_eval::const_eval::valtrees::slice_branches place=MPlaceTy { mplace: MemPlace { ptr: alloc1408, align: Align { pow2: 0 }, meta: Meta(0x000000000000002a) }, layout: TyAndLayout { ty: str, layout: Layout { fields: Array { stride: Size { raw: 1 }, count: 0 }, variants: Single { index: 0 }, abi: Aggregate { sized: false }, largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 0 } }, size: Size { raw: 0 } } } }
```

but in the bottom call:

```
│ │ │ │ │ │ ├─┐rustc_const_eval::const_eval::valtrees::fill_place_recursively place=MPlaceTy { mplace: MemPlace { ptr: alloc1411, align: Align { pow2: 0 }, meta: None }, layout: TyAndLayout { ty: u8, layout: Layout { fields: Primitive, variants: Single { index: 0 }, abi: Scalar(Initialized { value: Int(I8, false), valid_range: 0..=255 }), largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 0 } }, size: Size { raw: 1 } } } }, valtree=Leaf(0x66)
```

the write_immediate call fails with the error:

```
│ │ │ │ │ │ │ │ │ ├─┐rustc_const_eval::interpret::place::write_immediate_to_mplace_no_validate value=Scalar(0x66), dest=MPlaceTy { mplace: MemPlace { ptr: alloc1411, align: Align { pow2: 0 }, meta: None }, layout: TyAndLayout { ty: u8, layout: Layout { fields: Primitive, variants: Single { index: 0 }, abi: Scalar(Initialized { value: Int(I8, false), valid_range: 0..=255 }), largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 0 } }, size: Size { raw: 1 } } }
```

```
InterpErrorInfo(InterpErrorInfoInner { kind: memory access failed: alloc1411 has size 0, so pointer to 1 byte starting at offset 0 is out-of-bounds, backtrace: None })
```

Trying to figure out why this isn't working, but if one of you know what's going on, I'd appreciate your thoughts.

**Comment by b-naber** [comment]
Thanks for the reviews, addressed your comments.

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+            // create a place for the pointee
+            let mut pointee_place = create_pointee_place(&mut ecx, *inner_ty, valtree);
+            debug!(?pointee_place);
+
+            // insert elements of valtree into `place`
+            fill_place_recursively(&mut ecx, &mut pointee_place, valtree);
+            dump_place(&ecx, pointee_place.into());
+            intern_const_alloc_recursive(&mut ecx, InternKind::Constant, &pointee_place).unwrap();
+
+            let ref_place = pointee_place.to_ref(&tcx);
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:216]: Except for this line and the following one, this match arm is equal to the next one. Please merge them and run these two lines conditionally by matching on the type within the arm

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+                }
+                _ => (*place, branches, None),
+            };
+            debug!(?place_adjusted, ?branches);
+
+            // Create the places for the fields and fill them recursively
+            for (i, inner_valtree) in branches.iter().enumerate() {
+                debug!(?i, ?inner_valtree);
+
+                let mut place_inner = match *ty.kind() {
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:333]: Is this special casing needed? Instead of returning None for the variant index, you could return the zero variant id

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+                    ty::Adt(def, substs) if !def.is_enum() => {
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:334]: Why is this arm needed? It seems like it should just work to take the arm below

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 13]
```diff
@@ -0,0 +1,479 @@
...
+        | ty::RawPtr(_)
+        | ty::Str
+        | ty::Slice(_)
+        | ty::Dynamic(..) => bug!("no ValTree should have been created for type {:?}", ty.kind()),
+    }
+}
+
+// FIXME Needs a better/correct name
+#[instrument(skip(ecx), level = "debug")]
+fn fill_place_recursively<'tcx>(
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/valtrees.rs:262]: Maybe just leave a FIXME for this?

**Comment by oli-obk** [comment]
Idea for making sure this doesn't regress: add a call to the new logic behind an `if cfg!(debug_assertions)`, that should at least ensure that using it doesn't ICE.

**Review by oli-obk — APPROVED** [review]
Looks good to me in general now. From the logs in CI it seems there's still a bug in slice handling, but other than that this lgtm

**Comment by b-naber** [comment]
Is it ok if I run CI with `RUSTC_LOG=rustc_const_eval` or would that generate too much output? Alternatively maybe with select `println` statements?

**Comment by b-naber** [comment]
The bug has to originate in [this `mplace_field`](https://github.com/rust-lang/rust/blob/f7eae4e580d1e4db03800732421288c2ab557f1a/compiler/rustc_const_eval/src/const_eval/valtrees.rs#L333) call.

**Comment by oli-obk** [comment]
> Is it ok if I run CI with `RUSTC_LOG=rustc_const_eval` or would that generate too much output? Alternatively maybe with select `println` statements?

I think all you need to do locally is to run with `--stage 2` and the bug should reproduce.

I think we should figure out how to reproduce, ideally figure out the constant that is failing and add it to the ui test suite (assuming the test suite passes for you locally?)

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 14]
```diff
@@ -330,6 +331,13 @@ fn fill_place_recursively<'tcx>(
                 debug!(?i, ?inner_valtree);
 
                 let mut place_inner = match *ty.kind() {
+                    ty::Adt(def, substs) if !def.is_enum() => {
+                        let field = &def.variant(VariantIdx::from_usize(0)).fields[i];
+                        let field_ty = field.ty(tcx, substs);
+                        let projection_elem = ProjectionElem::Field(Field::from_usize(i), field_ty);
+
+                        ecx.mplace_projection(&place_adjusted, projection_elem).unwrap()
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:339]: This should be equivalent to mplace_field. I'm pretty sure all we do is call it inside `mplace_projection`

**Comment by b-naber** [comment]
[This](https://github.com/rust-lang/rust/blob/b759b2218649016cc40e82bdd6d958e2277ff6d7/library/core/src/ffi/c_str.rs#L87) is the struct that causes the ICE. There's a FIXME that says that `inner` shouldn't be unsized ^^

**Comment by oli-obk** [comment]
Oh, yea. I see the problem, but don't know the solution.

Unsized structs behind pointers are handled correctly during valtree creation, but going the other way only knows about str and slice as unsized types.

There are multiple ways to go about this, but I don't know the best one.

We could make references explicit in the valtree by always encoding them as a branch encoding either the thin pointers with one element or the wide pointers with two elements.

If we want to keep the way we encode valtrees (which I prefer), we'll need to find the unsized tail of types behind a reference. We can't just directly jump to the tail though, we need to walk it in lockstep with the valtree so we find the valtree branches of the unsized field.

We already have struct_tail_with_normalize which finds such a struct tail, but we have no way to walk it in lockstep with the valtree. So I think the right way to fix it is to add a closure argument to struct_tail_with_normalize that takes no arguments and has no return value. This closure could then be called after the match inside the for loop in struct_tail_with_normalize and we could supply a closure that grabs the last field of the valtree. This way we can keep getting that last field until we find the unsized tail type, which then must be either Str or Slice.

Does this make sense to you? I skipped a few impl steps needed, but that should be the gist of it

**Comment by b-naber** [comment]
It does make sense, though I don't understand why we need `struct_tail_with_normalize` to walk the type in lockstep with the valtree. Afaict there should be a branch for the unsized type. E.g. in 

```
struct Outer<'a, B: 'a + ?Sized> {
  first: u32,
  ref_unsized: &'a Foo<B>,
  second: &'a str,
}

struct Foo<B: ?Sized> {
  a: u32,
  b: B
}

struct Bar {
  c: u32,
  d: u32,
}

const ARR: Foo<[Bar; 2]> = Foo {
       a: 0,
       b: [Bar { c: 1, d: 2 }, Bar { c: 3, d: 4 }]
   };  

const SIZED: Outer<[Bar]> = Outer { first: 1, ref_unsized: &ARR, second: "Outer"};

fn main() {}
```

We get this valtree: `valtree=Branch([Leaf(0x00000001), Branch([Leaf(0x00000000), Branch([Branch([Leaf(0x00000001), Leaf(0x00000002)]), Branch([Leaf(0x00000003), Leaf(0x00000004)])])]), Branch([Leaf(0x4f), Leaf(0x75), Leaf(0x74), Leaf(0x65), Leaf(0x72)])])`, so when we get to the field `b` of `Foo` in the `fill_place_recursively` we have:

```
│ │ │ │ ├─0ms DEBUG rustc_const_eval::const_eval::valtrees i=1, inner_valtree=Branch([Branch([Leaf(0x00000001), Leaf(0x00000002)]), Branch([Leaf(0x00000003), Leaf(0x00000004)])])
│ │ │ │ ├─0ms DEBUG rustc_const_eval::const_eval::valtrees ty: Foo<[Bar]>, field: FieldDef { did: DefId(0:12 ~ dst[1ca3]::Foo::b), name: "b", vis: Restricted(DefId(0:0 ~ dst[1ca3])) }, field_ty: [Bar]
```

and can just use that branch. Am I missing something here?

Though I'm not sure what the correct `MPlace` to create here is. In the recursive call we're supposed to fill this place:

```
│ │ │ ├─┐rustc_const_eval::const_eval::valtrees::fill_place_recursively place=MPlaceTy { mplace: MemPlace { ptr: alloc16, align: Align { pow2: 2 }, meta: None }, layout: TyAndLayout { ty: Foo<[Bar]>, layout: Layout { fields: Arbitrary { offsets: [Size { raw: 0 }, Size { raw: 4 }], memory_index: [0, 1] }, variants: Single { index: 0 }, abi: Aggregate { sized: false }, largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 2 }, pref: Align { pow2: 3 } }, size: Size { raw: 4 } } } }, valtree=Branch([Leaf(0x00000000), Branch([Branch([Leaf(0x00000001), Leaf(0x00000002)]), Branch([Leaf(0x00000003), Leaf(0x00000004)])])])
```

Do we need to create an array for the place corresponding to `b` and use a relocation in the place we were supposed to fill?

**Comment by b-naber** [comment]
nvm I think you mean that we need to walk the type before the creation of the place and need the valtree to determine the size of the backing array in the place.

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/valtrees.rs:317]: The ICEs you are hitting in CI are because of this 😆 `fill_place_recursively` does not know about `FnDef`.

**Comment by b-naber** [comment]
@oli-obk Some tests involving statics still fail, this is due to `const_to_valtree` currently being called in the wrong place. I'm currently implementing the functionality for transitioning to using Valtrees in the type system (which will be part of a future PR though) and we won't encounter those errors then, since `const_to_valtree` will be called in the correct places. I'd suggest to run the test set including the call behind the debug_assertions cfg guard here once and then remove the call again.

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+
+            // Create a place for the underlying array
+            let len = valtree.unwrap_branch().len() as u64;
+            let arr_ty = tcx.mk_array(slice_ty, len as u64);
+            let place = create_mplace_from_layout(ecx, arr_ty);
+            debug!(?place);
+
+            place
+        }
+        ty::Adt(_, _) if !ty.is_sized(ecx.tcx, ty::ParamEnv::empty()) => {
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:213]: I think you can co down this arm for all unsized types ,even Str and Slice. Notethat tuples and other types can be unsized, too, so it's best to just use the `is_unsized` function instead of matching on they type

File: compiler/rustc_middle/src/ty/util.rs [hunk 16]
```diff
@@ -220,6 +220,10 @@ impl<'tcx> TyCtxt<'tcx> {
         self,
         mut ty: Ty<'tcx>,
         mut normalize: impl FnMut(Ty<'tcx>) -> Ty<'tcx>,
+        // This is a hack that is currently used to allow us to walk a ValTree
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/util.rs:223]: Not really a hack imo ^^

**Comment by b-naber** [comment]
Test suite results were as expected. Took the call of `valtree_to_const_value` out again.

**Comment by oli-obk** [comment]
While I was hoping we could leave it in, let's merge this and address the issue with statics in a separate PR

@bors r+

File: compiler/rustc_const_eval/src/interpret/place.rs [hunk 17]
```diff
@@ -115,6 +115,12 @@ impl<'tcx, Tag: Provenance> std::ops::Deref for MPlaceTy<'tcx, Tag> {
     }
 }
 
+impl<'tcx, Tag: Provenance> std::ops::DerefMut for MPlaceTy<'tcx, Tag> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.mplace
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/place.rs:120]: Not having mutable access to these fields was a deliberate decision. Why is this needed?

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 18]
```diff
@@ -0,0 +1,459 @@
...
+    ecx: &mut CompileTimeEvalContext<'tcx, 'tcx>,
+    ty: Ty<'tcx>,
+    valtree: ty::ValTree<'tcx>,
+) -> MPlaceTy<'tcx> {
+    let tcx = ecx.tcx.tcx;
+
+    if !ty.is_sized(ecx.tcx, ty::ParamEnv::empty()) {
+        // We need to create `Allocation`s for custom DSTs
+
+        let (unsized_inner_ty, num_elems) = get_info_on_unsized_field(ty, valtree, tcx);
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:200]: If we only need all this to get the size, then maybe it'd make sense to alternatively have a type for a `(ValTree, Option<Metadata>)`? That's often what our reference types end up looking like, so it probably makes sense here, too.

OTOH it is true that the valtree should already contain all the info we need, it's just a bit awkward to get it. Do we plan to ever have valtrees for `dyn Trait`?

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 19]
```diff
@@ -0,0 +1,459 @@
...
+                _ => create_mplace_from_layout(&mut ecx, ty),
+            };
+            debug!(?place);
+
+            fill_place_recursively(&mut ecx, &mut place, valtree);
+            dump_place(&ecx, place.into());
+            intern_const_alloc_recursive(&mut ecx, InternKind::Constant, &place).unwrap();
+
+            let const_val = match ty.kind() {
+                ty::Ref(_, _, _) => {
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:281]: Given these two `match` that special-case `Ref`, it might make sense to just make this a separate arm in the top-level `match`?

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 20]
```diff
@@ -0,0 +1,459 @@
...
+        | ty::RawPtr(_)
+        | ty::Str
+        | ty::Slice(_)
+        | ty::Dynamic(..) => bug!("no ValTree should have been created for type {:?}", ty.kind()),
+    }
+}
+
+// FIXME Needs a better/correct name
+#[instrument(skip(ecx), level = "debug")]
+fn fill_place_recursively<'tcx>(
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:318]: This is basically the destination passing style version of `valtree_to_const_value`. So what about something like `valtree_into_mplace`?

> I don't like that this duplicates logic from valtree_to_const_value

FWIW I am not surprised that something like this is required. It should not have a lot of duplication though; looks like that got improved since the comment has been written.

File: compiler/rustc_const_eval/src/interpret/place.rs [hunk 21]
```diff
@@ -487,7 +494,8 @@ where
     }
 
     /// Project into an mplace
-    pub(super) fn mplace_projection(
+    #[instrument(skip(self), level = "debug")]
+    pub(crate) fn mplace_projection(
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/place.rs:498]: Looks like this is not even used after all, and could hence remain private?

File: compiler/rustc_const_eval/src/interpret/place.rs (hunk 17 above)
Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/place.rs:120]: I think this is just leftover from previous refactorings. I missed that one :/

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs (hunk 18 above)
Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/valtrees.rs:200]: > Do we plan to ever have valtrees for `dyn Trait`?

Not at present.



> If we only need all this to get the size, then maybe it'd make sense to alternatively have a type for a `(ValTree, Option<Metadata>)`?

you mean an additional variant on `ValTree`? This is probably going in the same bucket as the string representation optimization that we'll look into. Until that happens (if ever) we could just add more convenience helpers

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs (hunk 19 above)
Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/valtrees.rs:281]: 😆 this was intentionally requested by me.

The three calls above and the `opt_to_const` were shared and the difference between the arms wasn't very obvious to me, with this change the difference is very obvious.

The first match is actually unnecessary, we can just unconditionally call `create_pointee_place` (though it should probably be renamed)

File: compiler/rustc_const_eval/src/interpret/place.rs (hunk 17 above)
Comment (b-naber) [reply compiler/rustc_const_eval/src/interpret/place.rs:120]: this is currently made use of here: https://github.com/b-naber/rust/blob/ef5f07256cfa4e1f43a18acb45e0d8108f8471a4/compiler/rustc_const_eval/src/const_eval/valtrees.rs#L227

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs (hunk 19 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/const_eval/valtrees.rs:281]: Fair.^^ I wouldn't even expect a lot of commonalities between these arms so having them into one seems a bit forced to me. :shrug:

File: compiler/rustc_const_eval/src/interpret/place.rs (hunk 17 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/place.rs:120]: I think it'd be better to construct a new `MPlaceTy` than change the existing one.
